game-logic.js

```diff
@@ -0,0 +1,367 @@
+// https://github.com/rocketacademy/basics-beat-that
+
+// 1. There are TWO players and players take turns.
+// 2. Each player roll TWO dice
+
+const PLAYER_ONE = 0;
+const PLAYER_TWO = 1;
+const LAST_PLAYER = PLAYER_TWO;
+const PLAYERS = 2;
+var CURRENT_PLAYER = 0;
+
+const PLAYER_WIN_COUNT = [0, 0];
+/**
+ * playerRolls
+ * An example outcome when both players throw dice
+ * playerRolls =  [
+ *                   [[1, 3]],
+ *                  [[6, 6]]
+ *                ];
+ */
+var playerRolls = [[], []];
+
+/**
+ * playersRollsOrderings
+ * An example outcome when both players throw dice
+ * playersRollsOrderings =   [
+ *                            [[1, 3],[3, 1]],
+ *                            [[6, 6]]
+ *                          ];
+ */
+var playersRollsOrderings = [[], []];
+
+/**
+ * playerOrderingChoice
+ * A example outcome when both players had chosen their ordering
+ * playerOrderingChoice =   [
+ *                            [[3, 1]],
+ *                            [[6, 6]]
+ *                          ];
+ */
+var playerOrderingChoice = [[], []];
+
+const ACTION_ROLL = 0;
+const ACTION_ORDER = 1;
+const ACTION_GAME_SET = 2;
+
+var CURRENT_ACTION = ACTION_ROLL;
+
+var PREV_ROUND_RESULT_DESCRIPTION = "";
+
+/** State Transition Functions */
+const resetGame = function () {
+  CURRENT_PLAYER = PLAYER_ONE;
+  playerRolls = [[], []];
+  playersRollsOrderings = [[], []];
+  playerOrderingChoice = [[], []];
+  PREV_ROUND_RESULT_DESCRIPTION = "";
+  return "Game reset";
+};
+const changeAction = function (ACTION) {
+  CURRENT_ACTION = ACTION;
+};
+const changePlayer = function () {
+  CURRENT_PLAYER = (CURRENT_PLAYER + 1) % PLAYERS;
+};
+
+/* --- Helper Functions --- */
+
+/* Dice motions */
+
+const rollDie = function () {
+  var face = 6;
+  return Math.floor(Math.random() * face) + 1;
+};
+
+// Permutations
+const generateOrderings = function (player, twoDiceValue) {
+  const [val1, val2] = twoDiceValue;
+  playersRollsOrderings[player].push([val1, val2]);
+  if (val2 !== val1) {
+    playersRollsOrderings[player].push([val2, val1]);
+  }
+};
+
+// How many choices does the player have?
+const getOrderingChoiceCount = function (player) {
+  return playersRollsOrderings[player].length;
+};
+
+// Check if index is in the orderings array of current player
+const isValidRollOrdering = function (PLAYER, index) {
+  if (
+    index === undefined ||
+    index === null ||
+    playersRollsOrderings[index] === undefined
+  ) {
+    return false;
+  }
```

Comment:
this prob can be reduced to just `playersRollsOrderings[index] === undefined` because if index is null or defined, then looking up the array index at null or undefined will return undefined anyway, making the condition true